sqlserver/datadog_checks/sqlserver/schemas.py

```diff
@@ -0,0 +1,401 @@
+                        "name": str
+                        "type": str
+                        "is_unique": bool
+                        "is_primary_key": bool
+                        "is_unique_constraint": bool
+                        "is_disabled": bool,
+                        "column_names": str
+                foreign_keys : list of foreign key dicts
+                    foreign_key
+                    key/value:
+                        "foreign_key_name": str
+                        "referencing_table": str
+                        "referencing_column": str
+                        "referenced_table": str
+                        "referenced_column": str
+                partitions: partition dict
+                    partition
+                    key/value:
+                        "partition_count": int
+        """
+        start_time = time.thread_time()
+        self._data_submitter.reset()
+        self._data_submitter.set_base_event_data(
+            self._check.resolved_hostname,
+            self._check.non_internal_tags,
+            self._check._config.cloud_metadata,
+            "{},{}".format(
+                self._check.static_info_cache.get(STATIC_INFO_VERSION, ""),
+                self._check.static_info_cache.get(STATIC_INFO_ENGINE_EDITION, ""),
+            ),
+        )
+
+        databases = self._check.get_databases()
+        db_infos = self._query_db_informations(databases)
+        self._data_submitter.store_db_infos(db_infos)
+
+        @tracked_method(agent_check_getter=agent_check_getter)
+        def fetch_schema_data(cursor, db_name):
+            schemas = self._query_schema_information(cursor)
+            for schema in schemas:
+                tables = self._get_tables(schema, cursor)
+                tables_chunks = list(get_list_chunks(tables, self.TABLES_CHUNK_SIZE))
+                for tables_chunk in tables_chunks:
+                    if time.thread_time() - start_time > self.MAX_EXECUTION_TIME:
+                        # TODO Report truncation to the backend
+                        self._log.warning(
+                            """Truncated data due to the effective execution time reaching {},
+                             stopped on db - {} on schema {}""".format(
+                                self.MAX_EXECUTION_TIME, db_name, schema["name"]
+                            )
+                        )
+                        raise StopIteration
+
+                    columns_count, tables_info = self._get_tables_data(tables_chunk, schema, cursor)
+
+                    self._data_submitter.store(db_name, schema, tables_info, columns_count)
+                    if self._data_submitter.columns_since_last_submit() > self.MAX_COLUMNS_PER_EVENT:
+                        self._data_submitter.submit()
+            self._data_submitter.submit()
+            return False
+
+        errors = self._check.do_for_databases(fetch_schema_data, self._check.get_databases())
```

Comment:
It's a bit weird to have this `do_for_databases` defined in the main `SqlServer` class. Is there any reason why we choose to loop through database inside the main check? could we have the schema collection log stay inside the `Schema` class?

Reply:
haha, well my idea was to refactor later code in sqlserver to use `do_for_databases ` in several places. Namely - 
[https://github.com/DataDog/integrations-core/blob/f176d94b085d92223ffb27fc2c12b1744062a0a2/sqlserver/datadog_checks/sqlserver/sqlserver.py#L728](url) and also there was one more place there was a bug that Jojo fixed, I'll need to find where it was but the idea is the same we need to iterate through DBs by USE if it's not Azure. But then I got several negative comments on `do_for_databases `  so I'd be fine to remove it in favour of a loop inside of schemas and then do a refactoring later. Dont know , what do you think ? Remove it and just do refactoring in the next PR  or keep it ?

Reply:
Ahh i missed this reply and added another comment about `do_for_databases`.  I feel the abstraction is slightly overkill and might have side effect in case we move things to a separate thread in future - pyodbc or adodbapi connection/cursor is not thread safe.

Reply:
yeah , came here cause wanted to send you the link. 

Well, fine ... I just think we should have a unified logic as this is a pattern. We will need it in at least 3 places, connection can be passed as an argument and do_for_databases can go in utils.

But we can do it later, so I'll move it to schemas.

Reply:
ps done